File: rcon4j-core/src/main/java/moe/mukjep/rcon4j/RconClient.java
package moe.mukjep.rcon4j;

import moe.mukjep.rcon4j.exceptions.AuthenticationException;

import java.io.IOException;
import java.io.OutputStream;
import java.net.Socket;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.util.Random;

public final class RconClient implements AutoCloseable {
    private final int requestId;

    private final Charset charset;

    private final Socket socket;

    public RconClient(String host, int port, String password) throws IOException, AuthenticationException {
        this(host, port, password, StandardCharsets.UTF_8);
    }

    public RconClient(String host, int port, String password, Charset charset) throws IOException, AuthenticationException {
        this(host, port, password, charset, new Random().nextInt(Integer.MAX_VALUE) + 1);
    }

    public RconClient(String host, int port, String password, Charset charset, int requestId) throws IOException, AuthenticationException {
        if (requestId < 0) {
            throw new IllegalArgumentException("requestId can't be negative");
        }

        this.requestId = requestId;
        this.charset = charset;
        this.socket = new Socket(host, port);
        RconPacket response = send(PacketTypes.SERVERDATA_AUTH, password.getBytes(charset));

        if (response.getId() == -1) {
            throw new AuthenticationException("Authentication failed");
        }
    }

    @Override
    public void close() throws IOException {
        socket.close();
    }

    public String command(String payload) throws IOException {
        if (payload == null || payload.isEmpty()) {
            throw new IllegalArgumentException("Payload can't be null or empty");
        }

        RconPacket response = send(PacketTypes.SERVERDATA_EXECCOMMAND, payload.getBytes(this.charset));

        if (response.getId() != this.requestId) {
            throw new IllegalStateException("Unexpected request id: " + response.getId());
        }

        return new String(response.getPayload(), this.charset);
    }

    private RconPacket send(int type, byte[] payload) throws IOException {
        OutputStream outputStream = socket.getOutputStream();
        outputStream.write(RconPacket.createPacket(requestId, type, payload));
        outputStream.flush();
        return RconPacket.fromStream(socket.getInputStream());
    }

}

File: rcon4j-core/src/main/kotlin/moe/mukjep/rcon4j/RconClient.kt
package moe.mukjep.rcon4j

import moe.mukjep.rcon4j.RconPacket.Companion.readRconPacket
import moe.mukjep.rcon4j.exceptions.AuthenticationException
import java.io.BufferedInputStream
import java.io.BufferedOutputStream
import java.io.IOException
import java.net.Socket
import java.nio.charset.Charset
import java.nio.charset.StandardCharsets
import java.util.*
import kotlin.random.Random

class RconClient @JvmOverloads @Throws(IOException::class, AuthenticationException::class) constructor(
    host: String,
    port: Int,
    password: String,
    private val charset: Charset = StandardCharsets.UTF_8,
    private val requestId: Int = Random.nextInt(1, Int.MAX_VALUE)
) : AutoCloseable {
    init {
        require(requestId >= 0) { "requestId can't be negative" }
    }

    private val socket = Socket(host, port)
    private val input = BufferedInputStream(socket.getInputStream())
    private val output = BufferedOutputStream(socket.getOutputStream())

    init {
        val response = send(PacketTypes.SERVERDATA_AUTH, password.toByteArray(charset))
        if (response.id == -1) {
            throw AuthenticationException("Authentication failed")
        }
    }

    @Throws(IOException::class)
    override fun close() {
        socket.close()
    }

    @Throws(IOException::class)
    fun command(payload: String): String {
        require(payload.isNotEmpty()) { "Payload can't be empty" }

        val response = send(PacketTypes.SERVERDATA_EXECCOMMAND, payload.toByteArray(charset))

        check(response.id == requestId) { "Unexpected request id: " + response.id }

        return response.payload.toString(charset)
    }

    @Throws(IOException::class)
    private fun send(type: Int, payload: ByteArray): RconPacket {
        output.write(RconPacket.bytes(requestId, type, payload))
        output.flush()
        return input.readRconPacket()
    }
}